xlators/storage/posix/src/posix-helpers.c

```diff
@@ -448,32 +449,26 @@ _posix_xattr_get_set(dict_t *xattr_req, char *key, data_t *data,
 
         /* file content request */
         req_size = data_to_uint64(data);
-        if (req_size >= filler->stbuf->ia_size) {
+        if (req_size >= filler->stbuf->ia_size && filler->stbuf->ia_size > 0) {
```

Comment:
I'm not sure if this is important, but this changes the behavior of this function. Previously, the key was created even if the data was 0 bytes. Now the key is not created if it's an empty file. I'm not sure if users of quick-read handle the same way a missing key than a key with empty data.

Reply:
I agree - I wasn't sure what's the point of quick read with zero data though. What's quick and what is it reading actually?

Reply:
If I'm not wrong, quick-read was an attempt to improve performance for small-file read workloads.

What it basically does is that it pre-reads the first 64 KiB (by default) of each file during lookup. So if nothing more than the first 64 KiB are required by the application, the read() operation can be directly served by quick-read, without any additional request to the brick.

If this is combined with open-behind and the application doesn't do anything else (i.e. just open, read, close), then both open() and read() requests are served locally without sending additional requests to the brick (though this has some problems (#3785) but not directly related to quick-read).

While we haven't seen clear indications that this could be a problem (at least I'm not aware of any bug reports that can be tracked to this), I think that this is an unsafe way to optimize reads. During lookups, writes could happen concurrently because there's no synchronization, so it's hard to determine which bricks contain healthy/up to date data directly from lookup request. AFR tries to do that and it seems it works well enough (at least we don't have reports). For disperse things are more difficult and actually it's not supported.

On the bad side, it could cause a significant increase of bandwidth and latency for lookups, which is specially bad if we are not going to take advantage of the pre-cached reads. It could also add another source of temporal inconsistency because of the lazy invalidation implementation we currently use for the caches. Additionally it can easily skip some features, like compression, encryption or anything else because xlators easily miss this special feature in lookups (actually only AFR handles this completely. EC and shard just disable it).

Reply:
I understand all the above - just unsure any of it matters for zero sized files - since I was the impression it's data that we read, not metadata.

Reply:
I would need to check the code, so I'm not sure, but it could happen that if the data is not present, a read is sent. However if data is present, even if it's of size 0, the read could not be sent and served directly from quick-read, which is its purpose.

Maybe it's irrelevant if other xlators don't care about this case, but since it's a change of behavior, I pointed it out.

I can't tell for sure if it's required or not.

Reply:
Removed that check.